python/hail/expr/aggregators/aggregators.py

```diff
@@ -49,47 +49,44 @@ def __init__(self):
                       ret_type=hail_type,
                       constructor_args=sequenceof(expr_any),
                       init_op_args=nullable(sequenceof(expr_any)),
-                      f=nullable(func_spec(1, expr_any)))
-    def __call__(self, name, aggregable, ret_type, constructor_args=(), init_op_args=None, f=None):
+                      seq_op_args=nullable(sequenceof(oneof(expr_any, func_spec(1, expr_any)))))
+    def __call__(self, name, aggregable, ret_type, constructor_args=(), init_op_args=None, seq_op_args=None):
         args = constructor_args if init_op_args is None else constructor_args + init_op_args
+        if seq_op_args is None:
+            seq_op_args = [lambda x: x]
         indices, aggregations = unify_all(aggregable, *args)
         if aggregations:
             raise ExpressionException('Cannot aggregate an already-aggregated expression')
         for a in args:
             _check_agg_bindings(a)
         _check_agg_bindings(aggregable)
 
+        uid = Env.get_uid()
+
         def get_type(expr):
             return expr.dtype
 
         def get_ir(expr):
             return expr._ir
 
-        def agg_sig(*seq_op_args):
+        def apply_seq_ops(seq_op_args, ref):
+            return [x(ref) if callable(x) else x for x in seq_op_args]
+
+        def agg_sig(applied_seq_ops):
             return AggSignature(name,
                                 list(map(get_type, constructor_args)),
                                 None if init_op_args is None else list(map(get_type, init_op_args)),
-                                list(map(get_type, seq_op_args)))
-
-        if name == "Count":
-            def make_seq_op(_):
-                return construct_expr(SeqOp(I32(0), [], agg_sig()), None)
-            seq_op = aggregable._transformations(aggregable, make_seq_op)
-            signature = agg_sig()
-        elif f is None:
-            def make_seq_op(agg):
-                return construct_expr(SeqOp(I32(0), [get_ir(agg)], agg_sig(aggregable)), None)
-            seq_op = aggregable._transformations(aggregable, make_seq_op)
-            signature = agg_sig(aggregable)
-        else:
-            def make_seq_op(agg):
-                uid = Env.get_uid()
-                ref = construct_variable(uid, get_type(agg), agg._indices)
-                result = f(ref)
-                body = Let(uid, get_ir(agg), get_ir(result))
-                return construct_expr(SeqOp(I32(0), [get_ir(agg), body], agg_sig(agg, result)), None)
-            seq_op = aggregable._transformations(aggregable, make_seq_op)
-            signature = agg_sig(aggregable, f(aggregable))
+                                list(map(get_type, applied_seq_ops)))
+
+        def make_seq_op(agg):
+            ref = construct_variable(uid, get_type(agg), agg._indices)
```

Comment:
I think you can move this into `apply_seq_ops` and eliminate the `ref` argument.